### ISSUE TITLE
Want to try porting this code as a Nodejs C++ Addon

### ISSUE THREAD

**Issue by DragonOsman** [nodejs/help#1590]
I have a web application that has a web server written in C++ using the Boost.Beast library (which relies on ASIO for the basic networking functionality and provides HTTP/S and WebSockets support on top of it), and also uses Jinja2Cpp and Nlohmann.JSON (former is a C++ implementation of Python's Jinja2 library, and the latter should be self-explanatory).  I used sync web server and client examples from the Beast repository on GitHub, and just added code to it to have it do what I wanted.  

The source code is on GitHub here: https://github.com/DragonOsman/currency_converter

I've been trying to build a Docker image of it so I can put it up on Heroku.  They require that the container be able to run correctly without root user access, so I had to add a normal user and I also installed `sudo` (rolled out my own image of Ubuntu with `sudo` and am pulling from it in the Dockerfile).  But for some reason I can't create directories in my home directory without using `sudo`.  I need a way to fix this.  

I'm asking for help here in understanding how the Node.js runtime and stuff works so I can port my C++ code to Node.js as a C++ Addon using V8 (though I also need to learn how to use V8 still).  As an alternative to the Docker+Heroku route if I can't get it to work.  I want to know how to expose my C++ app's networking functionality to Node.js and have the whole app run on the backend through Node.js.  Or if it'll be easier to port it to WebAssembly and use it through Node.js like that, then I'll do that (but, again, I need to know how to do this).

**Comment by DragonOsman** [comment]
It turns out that the problem with my Dockerfile build was that I was trying to run `mkdir` in `~/app`.  When I used my home directory itself as the working directory and ran `mkdir` in there, it worked without me needing to use `sudo`.  

I still want to know how to do this with V8 and/or WebAssembly, though, so some help would be appreciated (and yes, I want to fully understand how the Node.js runtime works so I can port ASIO and Beast to WebAssembly and use the networking functionality from my app in Node.js--and instead of running my code on the frontend as an ES module, I want to run it on the backend as a Node module since it's a backend application).

**Comment by ORESoftware** [comment]
@DragonOsman glad to help, but I think we need to pinpoint a specific question you have, sounds too general to me and I doubt anyone would know where to start to help.

**Comment by DragonOsman** [comment]
Need to make correction: not WebAssembly repo but the `emsdk` repo.  Sorry about that.

**Comment by DragonOsman** [comment]
I'm having errors with port binding when trying to deploy the app to Heroku as a Docker container.  They need to bind a port using the $PORT environment and I'm not sure if I'm doing it right.  And I ask them either because I have a free account due to which I can't open support tickets.  So I think I'll try with Node.js C++ Addons or WebAssembly (whichever is easier) after all.  

Please note that I do want to port the entire code, including the networking functionality.  Where d I need to read to understand how the Node.js runtime and everything works so that I can do this?  I need tutorials on V8 as well.  And also how to get `cwrap` and `ccall` to work when using Emscripten when porting to WebAssembly.  Thanks.

**Comment by DragonOsman** [comment]
My code uses synchronous networking functions, and as I understand it, Node.js and JavaScript require the asynchronous versions?  And I also read in the WebAssembly documentation that some functions in the C/C++ code might get thrown away as "dead code" by the optimizer.  Will this definitely happen to `main`?  Do I need to use `ccall` on `main`?  What about my other functions, since I need all of the functions in my code to be called as needed?

I was told that Boost.ASIO doesn't support WebAssembly, so I'll have to port it.  I'd like help on that as well.  Boost.Beast uses Boost.ASIO, and I use Beast for the HTTP stuff in the code.  I may have to port both ASIO and Beast.  https://github.com/boostorg/asio and https://github.com/boostorg/beast .  I used the synchronous HTTP server and client examples from the Beast repository in my code, and added the code I needed to add.  I'll try to port it to WebAssembly first.  If that's too hard, I'll fall back to V8 and C++ Addons.
